lib/vector-buffers/src/variants/disk_v2/v1_migration.rs

```diff
@@ -90,14 +87,12 @@ where
             )
         })?;
 
-        src_acker.ack(old_record_event_count);
```

Comment:
Hmm, this could get tricky because I believe we'd positively acknowledge the records as soon as `write_record` completes... even if there was an error. This obviously won't matter once `disk_v1` is gone but I'm not 100% sure it'll be gone before 0.24.0 is released.

I wonder if we would want to take the finalizers manually, and then manually acknowledge only once we've gotten past the flush. Although I suppose we're still at the mercy of the default drop behavior if we do an early return on `write_record`/`flush`.... hmmm.

Reply:
I can take a shot at handling taking the finalizers and updating their state on error, but at this point the delivery status is being ignored, so it will still always count as acknowledged even for error, which doesn't actually help the situation.

Reply:
Right, yeah, the "success by default" sort of hurts us here without a more complex mechanism to invert the logic.

One thought is that I _think_ you could extract each record's finalizers, and then wrap them in a `ManuallyDrop<T>`. If we succeed at writing and flushing the record, we could then manually drop the finalizers via `drop(wrapped_finalizers.into_inner())`: this would yield the current behavior of successfully finalizing the events when `EventFinalizer` is dropped. If we returned early due to an error, though, then we would never call that drop glue.

At best, we would leak some small amount of memory -- the `Vec` of finalizers? an `Arc` or something? -- but at that point, we're already exiting the program due to the early return, so I'm not sure it matters so long as we do not positively acknowledge the record we've just read out of the old buffer.

Reply:
`ManuallyDrop` is a great idea, I had forgotten about that. AFAICT we would only need to hang on to them until the `flush`, so yes we would only leak the finalizers from one record worth, which would end up being just one plus the `Vec`. The implementation is proving to be a bit of a nuisance since it means adding a `Finalizable` bound in a bunch of new places, which impacts tests, but it's all doable.

Reply:
Done in f4924ccb1